poc-07.py

```diff
@@ -167,31 +173,91 @@ def update(self, theirMessages):
     def getChanges(self):
         """Explore our messages. Only return changes since previous sync."""
 
-        changedMessages = Messages() # Collection of new, deleted and updated messages.
+        # Collection of new, deleted and updated messages.
+        changedMessages = Messages()
         messages = self.driver.search() # Would be async.
-        stateMessages = self.state.search() # Would be async.
+        ourStateMessages = self.state.search() # Would be async.
+        theirStateMessages = self.theirState.search() # Would be async
 
         #TODO: we have to read both states to know what to sync. The current
         # implementation is wrong.
-        for message in self.theirState.search():
-            if message not in stateMessages:
-                stateMessages.append(message)
+
+        for theirMessage in theirStateMessages:
+            if theirMessage not in ourStateMessages:
+                ourStateMessages.append(theirMessage)
 
         for message in messages:
-            if message not in stateMessages:
-                # Missing in the other side.
+            if message not in ourStateMessages:
```

Comment:
Nitpicking: most of the work is done on `ourStateMessages`. Hence, using this variable name is quite misleading.
I think it's better to keep `stateMessages`.